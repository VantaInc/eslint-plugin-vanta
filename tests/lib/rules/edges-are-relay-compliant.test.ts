import {
  GraphQLInvalidTestCase,
  GraphQLValidTestCase,
  ParserOptions,
} from "@graphql-eslint/eslint-plugin";

import { GraphQLRuleTester } from "@graphql-eslint/eslint-plugin";
import rule from "../../../lib/rules/edges-are-relay-compliant";

const createValid = (schema: string): GraphQLValidTestCase<any[]> => {
  return { code: schema };
};
const createInvalid = (
  schema: string,
  errors: number = 1
): GraphQLInvalidTestCase<any[]> => {
  return {
    code: schema,
    errors: errors,
  };
};

const ruleTester = new GraphQLRuleTester();

const valid1 = /* GraphQL */ `
  type T1 {
    a: String
  }

  type Edge {
    node: T1
    cursor: String
  }
`;

const valid2 = /* GraphQL */ `
  type T1 {
    a: String
  }

  type Edge {
    node: T1!
    cursor: String
  }
`;

const valid3 = /* GraphQL */ `
  type osquery {
    active: Boolean!
    domainId: ID
    hostIdentifier: String!
    id: ID!
    node_key: String!
    platform: String!
  }

  type EndpointEdge {
    cursor: String!
    node: osquery!
  }
`;

const invalid1 = /* GraphQL */ `
  type MissingNodeEdge {
    notNode: String
    cursor: String
  }
`;

const invalid2 = /* GraphQL */ `
  type ListNodeEdge {
    node: [String]
    cursor: String
  }
`;

const invalid3 = /* GraphQL */ `
  type NonNullListEdge {
    node: [String]!
    cursor: String
  }
`;

const invalid4 = /* GraphQL */ `
  type MissingCursorEdge {
    node: [String]!
  }
`;

const invalid5 = /* GraphQL */ `
  type NotStringCursorEdge {
    node: String!
    edge: Int
  }
`;

ruleTester.runGraphQLTests("edges-are-relay-compliant", rule, {
  valid: [createValid(valid1), createValid(valid2), createValid(valid3)],
  invalid: [
    createInvalid(invalid1),
    createInvalid(invalid2),
    createInvalid(invalid3),
    createInvalid(invalid4, 2),
    createInvalid(invalid5),
  ],
});
