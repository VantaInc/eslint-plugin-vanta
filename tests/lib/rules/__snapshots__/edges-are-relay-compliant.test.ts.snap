// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Invalid #1 1`] = `
"#### ⌨️ Code

      1 |   type MissingNodeEdge {
      2 |     notNode: String
      3 |     cursor: String
      4 |   }

#### ❌ Error

    > 1 |   type MissingNodeEdge {
        |   ^^^^^^^^^^^^^^^^^^^^^^
    > 2 |     notNode: String
        | ^^^^^^^^^^^^^^^^^^^
    > 3 |     cursor: String
        | ^^^^^^^^^^^^^^^^^^^
    > 4 |   }
        | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ Edges must contain a node field"
`;

exports[`Invalid #2 1`] = `
"#### ⌨️ Code

      1 |   type ListNodeEdge {
      2 |     node: [String]
      3 |     cursor: String
      4 |   }

#### ❌ Error

    > 1 |   type ListNodeEdge {
        |   ^^^^^^^^^^^^^^^^^^^
    > 2 |     node: [String]
        | ^^^^^^^^^^^^^^^^^^
    > 3 |     cursor: String
        | ^^^^^^^^^^^^^^^^^^
    > 4 |   }
        | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ Node field cannot be a list"
`;

exports[`Invalid #3 1`] = `
"#### ⌨️ Code

      1 |   type NonNullListEdge {
      2 |     node: [String]!
      3 |     cursor: String
      4 |   }

#### ❌ Error

    > 1 |   type NonNullListEdge {
        |   ^^^^^^^^^^^^^^^^^^^^^^
    > 2 |     node: [String]!
        | ^^^^^^^^^^^^^^^^^^^
    > 3 |     cursor: String
        | ^^^^^^^^^^^^^^^^^^^
    > 4 |   }
        | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ Node field cannot be a list"
`;

exports[`Invalid #4 1`] = `
"#### ⌨️ Code

      1 |   type MissingCursorEdge {
      2 |     node: [String]!
      3 |   }

#### ❌ Error 1/2

    > 1 |   type MissingCursorEdge {
        |   ^^^^^^^^^^^^^^^^^^^^^^^^
    > 2 |     node: [String]!
        | ^^^^^^^^^^^^^^^^^^^
    > 3 |   }
        | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ Node field cannot be a list

#### ❌ Error 2/2

    > 1 |   type MissingCursorEdge {
        |   ^^^^^^^^^^^^^^^^^^^^^^^^
    > 2 |     node: [String]!
        | ^^^^^^^^^^^^^^^^^^^
    > 3 |   }
        | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ Edges must contain a cursor field"
`;

exports[`Invalid #5 1`] = `
"#### ⌨️ Code

      1 |   type NotStringCursorEdge {
      2 |     node: String!
      3 |     edge: Int
      4 |   }

#### ❌ Error

    > 1 |   type NotStringCursorEdge {
        |   ^^^^^^^^^^^^^^^^^^^^^^^^^^
    > 2 |     node: String!
        | ^^^^^^^^^^^^^^^^^
    > 3 |     edge: Int
        | ^^^^^^^^^^^^^^^^^
    > 4 |   }
        | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ Edges must contain a cursor field"
`;
